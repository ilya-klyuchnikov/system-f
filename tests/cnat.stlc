;; check that fix propagates the type information inside
(let ((int->nat2 (fix f _
                   (lam (count-from)
                     (lam (s z)
                       (if0 count-from
                         z
                         (s ((f (- count-from 1)) s z))))))
                 (int -> (all (a) ((a -> a) a -> a)))))
  int->nat2)

;; ditto
(let ((+ (lam (n m) (- n (- 0 m)))
         (int int -> int))
      (int->nat1 (lam (n)
                   (Lam (a)
                     ((fix f (int -> (a -> a) a -> a)
                        (lam (count-from)
                          (lam (s z)
                            (if0 count-from
                              z
                              (s ((f (- count-from 1)) s z))))))
                      n)))
                 (int -> (all (a) ((a -> a) a -> a))))
      (int->nat2 (fix f _
                   (lam (count-from)
                     (lam (s z)
                       (if0 count-from
                         z
                         (s ((f (- count-from 1)) s z))))))
                 (int -> (all (a) ((a -> a) a -> a)))))
  (+ ((int->nat1 5)
      (lam ((acc int)) (+ acc 100))
      3)
     ((int->nat2 2)
      (lam ((acc int)) (+ acc 10))
      4)))
