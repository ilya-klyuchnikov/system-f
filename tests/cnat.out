>  : (-> int (all (a) (-> (-> a a) a a)))
-> #<function>
>  : int
-> 527
> 